File: scripts/dev/checksum/calculate.py
"""文件哈希计算实现"""
import hashlib
import sys
from pathlib import Path

# 添加父目录到路径，以便导入 utils
sys.path.insert(0, str(Path(__file__).parent.parent))
from utils.logger import log_info, log_success, log_error


def calculate(args) -> str:
    """计算文件 SHA256 哈希"""
    file_path = Path(args.file)

    if not file_path.exists():
        log_error(f"File not found: {args.file}")
        sys.exit(1)

    # 读取文件并计算哈希（使用标准库 hashlib）
    sha256_hash = hashlib.sha256()
    try:
        with open(file_path, 'rb') as f:
            # 分块读取，避免大文件占用过多内存
            for chunk in iter(lambda: f.read(4096), b''):
                sha256_hash.update(chunk)
    except IOError as e:
        log_error(f"Failed to read file: {e}")
        sys.exit(1)

    hash_hex = sha256_hash.hexdigest()

    log_info(f"File: {args.file}")
    log_info(f"SHA256: {hash_hex}")

    # 输出到文件或 stdout
    if args.output:
        output_path = Path(args.output)
        try:
            output_path.write_text(hash_hex)
            log_success(f"Hash written to: {args.output}")
        except IOError as e:
            log_error(f"Failed to write hash to {args.output}: {e}")
            sys.exit(1)
    else:
        # 直接输出哈希值（用于管道）
        print(hash_hex)

    return hash_hex


def main():
    """CLI 入口（可以直接运行此脚本）"""
    import argparse

    parser = argparse.ArgumentParser(
        prog='calculate',
        description='Calculate file SHA256 checksum'
    )
    parser.add_argument('--file', required=True, help='File path')
    parser.add_argument('--output', help='Output file path (optional)')

    args = parser.parse_args()
    calculate(args)


if __name__ == '__main__':
    main()

File: src/commands/dev/checksum/calculate.rs
//! 文件哈希计算实现

use crate::{log_info, log_success};
use color_eyre::{eyre::WrapErr, Result};
use sha2::{Digest, Sha256};
use std::fs;
use std::path::Path;

/// 文件哈希计算命令
pub struct ChecksumCalculateCommand {
    file: String,
    output: Option<String>,
}

impl ChecksumCalculateCommand {
    /// 创建新的文件哈希计算命令
    pub fn new(file: String, output: Option<String>) -> Self {
        Self { file, output }
    }

    /// 计算文件哈希
    pub fn calculate(&self) -> Result<String> {
        let file_path = Path::new(&self.file);

        if !file_path.exists() {
            return Err(color_eyre::eyre::eyre!("File not found: {}", self.file));
        }

        // 读取文件内容
        let file_content =
            fs::read(file_path).wrap_err_with(|| format!("Failed to read file: {}", self.file))?;

        // 计算 SHA256 哈希
        let mut hasher = Sha256::new();
        hasher.update(&file_content);
        let hash = hasher.finalize();
        let hash_hex = format!("{:x}", hash);

        log_info!("📄 File: {}", self.file);
        log_info!("🔐 SHA256: {}", hash_hex);

        // 输出到文件或标准输出
        if let Some(ref output_path) = self.output {
            fs::write(output_path, &hash_hex)
                .wrap_err_with(|| format!("Failed to write hash to: {}", output_path))?;
            log_success!("Hash written to: {}", output_path);
        }

        Ok(hash_hex)
    }
}
